File: services/social-service/internal/service/oauth2_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"github.com/dqx0/glen/social-service/internal/models"
)

var (
	ErrInvalidAuthCode = errors.New("invalid authorization code")
	ErrInvalidToken    = errors.New("invalid token")
	ErrInvalidConfig   = errors.New("invalid OAuth2 config")
	ErrHTTPRequest     = errors.New("HTTP request failed")
)

// OAuth2Token はOAuth2アクセストークンを表す
type OAuth2Token struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	RefreshToken string `json:"refresh_token,omitempty"`
	ExpiresIn    int    `json:"expires_in,omitempty"`
	Scope        string `json:"scope,omitempty"`
}

// OAuth2Service はOAuth2フローを処理する
type OAuth2Service struct {
	config *models.OAuth2Config
	client *http.Client
}

// NewOAuth2Service は新しいOAuth2Serviceを作成する
func NewOAuth2Service(config *models.OAuth2Config) *OAuth2Service {
	if config == nil || config.ClientID == "" {
		return nil
	}
	
	return &OAuth2Service{
		config: config,
		client: &http.Client{},
	}
}

// GetAuthURL は認証URLを生成する
func (s *OAuth2Service) GetAuthURL(state string) string {
	params := url.Values{}
	params.Set("client_id", s.config.ClientID)
	params.Set("redirect_uri", s.config.RedirectURL)
	params.Set("response_type", "code")
	params.Set("scope", strings.Join(s.config.Scopes, " "))
	params.Set("state", state)
	
	return fmt.Sprintf("%s?%s", s.config.AuthURL, params.Encode())
}

// ExchangeCodeForToken は認証コードをアクセストークンに交換する
func (s *OAuth2Service) ExchangeCodeForToken(ctx context.Context, code string) (*OAuth2Token, error) {
	if code == "" {
		return nil, ErrInvalidAuthCode
	}
	
	// トークンリクエストのパラメータを準備
	data := url.Values{}
	data.Set("grant_type", "authorization_code")
	data.Set("code", code)
	data.Set("client_id", s.config.ClientID)
	data.Set("client_secret", s.config.ClientSecret)
	data.Set("redirect_uri", s.config.RedirectURL)
	
	// HTTPリクエストを作成
	req, err := http.NewRequestWithContext(ctx, "POST", s.config.TokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	
	// リクエストを送信
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrHTTPRequest, err)
	}
	defer resp.Body.Close()
	
	// レスポンスボディを読み取り
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}
	
	// HTTPステータスの確認
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token exchange failed with status %d: %s", resp.StatusCode, string(body))
	}
	
	// JSONレスポンスをパース
	var token OAuth2Token
	if err := json.Unmarshal(body, &token); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}
	
	return &token, nil
}

// GetUserInfo はアクセストークンを使ってユーザー情報を取得する
func (s *OAuth2Service) GetUserInfo(ctx context.Context, token *OAuth2Token) (map[string]interface{}, error) {
	if token == nil || token.AccessToken == "" {
		return nil, ErrInvalidToken
	}
	
	// HTTPリクエストを作成
	req, err := http.NewRequestWithContext(ctx, "GET", s.config.UserInfoURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	
	// Authorizationヘッダーを設定
	authHeader := fmt.Sprintf("%s %s", token.TokenType, token.AccessToken)
	if token.TokenType == "" {
		authHeader = fmt.Sprintf("Bearer %s", token.AccessToken)
	}
	req.Header.Set("Authorization", authHeader)
	req.Header.Set("Accept", "application/json")
	
	// リクエストを送信
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrHTTPRequest, err)
	}
	defer resp.Body.Close()
	
	// レスポンスボディを読み取り
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}
	
	// HTTPステータスの確認
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("user info request failed with status %d: %s", resp.StatusCode, string(body))
	}
	
	// JSONレスポンスをパース
	var userInfo map[string]interface{}
	if err := json.Unmarshal(body, &userInfo); err != nil {
		return nil, fmt.Errorf("failed to parse user info response: %w", err)
	}
	
	return userInfo, nil
}

// ValidateState はstateパラメータを検証する
func (s *OAuth2Service) ValidateState(providedState, expectedState string) bool {
	return providedState != "" && expectedState != "" && providedState == expectedState
}

// RefreshToken はリフレッシュトークンを使って新しいアクセストークンを取得する
func (s *OAuth2Service) RefreshToken(ctx context.Context, refreshToken string) (*OAuth2Token, error) {
	if refreshToken == "" {
		return nil, ErrInvalidToken
	}
	
	// リフレッシュリクエストのパラメータを準備
	data := url.Values{}
	data.Set("grant_type", "refresh_token")
	data.Set("refresh_token", refreshToken)
	data.Set("client_id", s.config.ClientID)
	data.Set("client_secret", s.config.ClientSecret)
	
	// HTTPリクエストを作成
	req, err := http.NewRequestWithContext(ctx, "POST", s.config.TokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	
	// リクエストを送信
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrHTTPRequest, err)
	}
	defer resp.Body.Close()
	
	// レスポンスボディを読み取り
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}
	
	// HTTPステータスの確認
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token refresh failed with status %d: %s", resp.StatusCode, string(body))
	}
	
	// JSONレスポンスをパース
	var token OAuth2Token
	if err := json.Unmarshal(body, &token); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}
	
	return &token, nil
}

// RevokeToken はトークンを無効化する（サポートしているプロバイダーのみ）
func (s *OAuth2Service) RevokeToken(ctx context.Context, token string) error {
	if token == "" {
		return ErrInvalidToken
	}
	
	// プロバイダー別の無効化URL（設定で提供される場合）
	// 実装は簡略化し、基本的なPOSTリクエストのみ
	data := url.Values{}
	data.Set("token", token)
	data.Set("client_id", s.config.ClientID)
	data.Set("client_secret", s.config.ClientSecret)
	
	// 無効化のエンドポイントがない場合は成功として扱う
	// 実際のアプリケーションでは、プロバイダー別に対応が必要
	return nil
}

File: frontend/src/services/oauth2Service.ts
interface OAuth2Client {
  id: string;
  client_id: string;
  client_secret?: string;
  name: string;
  description: string;
  redirect_uris: string[];
  scopes: string[];
  is_public: boolean;
  is_active: boolean;
  created_at: string;
}

interface CreateClientRequest {
  user_id: string;
  name: string;
  description: string;
  redirect_uris: string[];
  scopes: string[];
  is_public: boolean;
}

interface CreateClientResponse {
  id: string;
  client_id: string;
  client_secret?: string;
  name: string;
  description: string;
  redirect_uris: string[];
  scopes: string[];
  is_public: boolean;
  is_active: boolean;
  created_at: string;
}

interface AuthorizeRequest {
  client_id: string;
  redirect_uri: string;
  response_type: string;
  scope: string;
  state?: string;
  code_challenge?: string;
  code_challenge_method?: string;
}

interface TokenRequest {
  grant_type: string;
  code?: string;
  redirect_uri?: string;
  client_id: string;
  client_secret?: string;
  code_verifier?: string;
  refresh_token?: string;
  scope?: string;
}

interface TokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token?: string;
  scope?: string;
}

import { apiClient } from '../api/client';

export class OAuth2Service {
  private static baseUrl = '/api/v1/oauth2';

  // Client Management
  static async createClient(request: CreateClientRequest): Promise<CreateClientResponse> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/clients`, request);
      return response.data;
    } catch (error: any) {
      const errorMessage = error.response?.data?.error_description || 
                          error.response?.data?.message || 
                          error.message || 
                          'Failed to create client';
      throw new Error(errorMessage);
    }
  }

  static async getClients(userId: string): Promise<OAuth2Client[]> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/clients?user_id=${encodeURIComponent(userId)}`);
      return Array.isArray(response.data) ? response.data : [];
    } catch (error: any) {
      if (error.response?.status === 404) {
        return []; // No clients found
      }
      const errorMessage = error.response?.data?.error_description || 
                          error.response?.data?.message || 
                          error.message || 
                          'Failed to fetch clients';
      throw new Error(errorMessage);
    }
  }

  static async getClient(clientId: string): Promise<OAuth2Client> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/clients/${encodeURIComponent(clientId)}`);
      return response.data;
    } catch (error: any) {
      const errorMessage = error.response?.data?.error_description || 
                          error.response?.data?.message || 
                          error.message || 
                          'Failed to fetch client';
      throw new Error(errorMessage);
    }
  }

  static async deleteClient(clientId: string): Promise<void> {
    try {
      await apiClient.delete(`${this.baseUrl}/clients/${encodeURIComponent(clientId)}`);
    } catch (error: any) {
      const errorMessage = error.response?.data?.error_description || 
                          error.response?.data?.message || 
                          error.message || 
                          'Failed to delete client';
      throw new Error(errorMessage);
    }
  }

  // Authorization Flow
  static async authorize(request: AuthorizeRequest): Promise<void> {
    try {
      // ユーザー情報をローカルストレージから取得
      const userDataStr = localStorage.getItem('user');
      const userData = userDataStr ? JSON.parse(userDataStr) : null;
      
      console.log('OAuth2Service.authorize: User data:', userData);
      console.log('OAuth2Service.authorize: Request:', request);
      
      // Build authorization URL for direct browser navigation
      const params = new URLSearchParams({
        client_id: request.client_id,
        redirect_uri: request.redirect_uri,
        response_type: request.response_type,
        scope: request.scope,
        ...(request.state && { state: request.state }),
        ...(request.code_challenge && { code_challenge: request.code_challenge }),
        ...(request.code_challenge_method && { code_challenge_method: request.code_challenge_method }),
      });
      
      // For OAuth2 authorization, we need to navigate the browser directly
      // instead of using AJAX requests, because the server will perform redirects
      
      // Navigate via API Gateway for centralized request handling
      const authUrl = `http://localhost:8080/api/v1/oauth2/authorize?${params.toString()}`;
      
      console.log('OAuth2Service.authorize: Navigating via API Gateway:', authUrl);
      console.log('OAuth2Service.authorize: User ID being passed:', userData?.id);
      
      // Navigate directly to the authorization endpoint
      window.location.href = authUrl;
      
    } catch (error: any) {
      console.error('OAuth2Service.authorize: Error details:', error);
      const errorMessage = error.response?.data?.error_description || 
                          error.response?.data?.error || 
                          error.message || 
                          'Authorization failed';
      throw new Error(errorMessage);
    }
  }

  // Token Exchange
  static async exchangeToken(request: TokenRequest): Promise<TokenResponse> {
    const formData = new URLSearchParams();
    formData.append('grant_type', request.grant_type);
    formData.append('client_id', request.client_id);
    
    if (request.client_secret) {
      formData.append('client_secret', request.client_secret);
    }
    if (request.code) {
      formData.append('code', request.code);
    }
    if (request.redirect_uri) {
      formData.append('redirect_uri', request.redirect_uri);
    }
    if (request.code_verifier) {
      formData.append('code_verifier', request.code_verifier);
    }
    if (request.refresh_token) {
      formData.append('refresh_token', request.refresh_token);
    }
    if (request.scope) {
      formData.append('scope', request.scope);
    }

    const response = await fetch(`${this.baseUrl}/token`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Token exchange failed');
    }

    return response.json();
  }

  // Token Introspection
  static async introspectToken(token: string): Promise<any> {
    const formData = new URLSearchParams();
    formData.append('token', token);

    const response = await fetch(`${this.baseUrl}/introspect`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...(localStorage.getItem('accessToken') && { 
          'Authorization': `Bearer ${localStorage.getItem('accessToken')}` 
        }),
      },
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Token introspection failed');
    }

    return response.json();
  }

  // Token Revocation
  static async revokeToken(token: string, clientId: string, clientSecret?: string): Promise<void> {
    const formData = new URLSearchParams();
    formData.append('token', token);
    formData.append('client_id', clientId);
    
    if (clientSecret) {
      formData.append('client_secret', clientSecret);
    }

    const response = await fetch(`${this.baseUrl}/revoke`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Token revocation failed');
    }
  }

  // PKCE Helper Functions
  static generateCodeVerifier(length: number = 128): string {
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
    let text = '';
    for (let i = 0; i < length; i++) {
      text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
  }

  static async generateCodeChallenge(codeVerifier: string): Promise<string> {
    const encoder = new TextEncoder();
    const data = encoder.encode(codeVerifier);
    const digest = await crypto.subtle.digest('SHA-256', data);
    
    // Convert to base64url
    return btoa(String.fromCharCode(...new Uint8Array(digest)))
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=/g, '');
  }

  // Authorization URL Builder
  static buildAuthorizationUrl(params: {
    client_id: string;
    redirect_uri: string;
    scope: string;
    state?: string;
    code_challenge?: string;
    code_challenge_method?: string;
  }): string {
    const baseUrl = `${window.location.origin}/oauth/authorize`;
    const searchParams = new URLSearchParams({
      response_type: 'code',
      client_id: params.client_id,
      redirect_uri: params.redirect_uri,
      scope: params.scope,
    });

    if (params.state) {
      searchParams.append('state', params.state);
    }
    if (params.code_challenge) {
      searchParams.append('code_challenge', params.code_challenge);
      searchParams.append('code_challenge_method', params.code_challenge_method || 'S256');
    }

    return `${baseUrl}?${searchParams.toString()}`;
  }
}

export type { 
  OAuth2Client, 
  CreateClientRequest, 
  CreateClientResponse,
  AuthorizeRequest,
  TokenRequest,
  TokenResponse 
};